#Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [2]:
import pandas as pd
df = pd.read_csv('/content/drive/MyDrive/Курс Дата аналітик /Data/supermarket_sales.csv')

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [3]:
df.groupby('Branch')['gross income'].sum().sort_values(ascending = False)[:1]

Branch
C    5265.1765
Name: gross income, dtype: float64

Найприбутковішим є філіал С.

2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [7]:
branch_income = df.groupby(['Branch','City'])['gross income'].sum().sort_values(ascending=False)[:1]
branch_income

Branch  City     
C       Naypyitaw    5265.1765
Name: gross income, dtype: float64

В місті Naypyitaw	знаходиться найприбутковіший філіал. Проте різниця між містами незначна, тому не впевнена, що варто відкривати там ще один магазин. Можливо це навпаки гірше вплине на прибуток в цьому місті (затрати будуть на +1 магазин).
Думаю треба дослідити наскільки часто це місто та цей філіал лідирують.

3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [ ]:
invoices_by_category = df.pivot_table(index = 'Branch', columns = 'Product line', values = 'Invoice ID', aggfunc = 'count')
invoices_by_category

Product line  Electronic accessories  Fashion accessories  Food and beverages  \
Branch                                                                          
A                                 60                   51                  58   
B                                 55                   62                  50   
C                                 55                   65                  66   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
Branch                                                                  
A                            47                  65                 59  
B                            53                  50                 62  
C                            52                  45                 45

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1rueAdko6S3UxIHGtojetTxlES-EyM6Yb)

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [ ]:
invoices_by_category['Electronic accessories'].max() #найбільше інвойсів

60

In [ ]:
invoices_by_category['Electronic accessories'].idxmax() #в якому філіалі

'A'

4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


**Питання:**

1. Який максимальний,мінімальний та середній рейтинг ставлять покупці в місті `"Naypyitaw"`?

2. Де найнижча `"cogs"`?

3. Якого `"Customer type"` більше в місті `"Naypyitaw"`?

**1 питання**

Який максимальний, мінімальний та середній рейтинг ставлять покупці в місті `"Naypyitaw"`?

In [ ]:
df['Rating'][df['City'] == 'Naypyitaw'].max() #максимальний рейтинг

10.0

In [ ]:
df['Rating'][df['City'] == 'Naypyitaw'].min() #мінімальний рейтинг

4.0

In [ ]:
df['Rating'][df['City'] == 'Naypyitaw'].mean() #середній рейтинг

7.072865853658537

In [ ]:
naypyitaw_ratings = df['Rating'][df['City'] == 'Naypyitaw']
with pd.option_context("display.precision", 2):
  display(naypyitaw_ratings.describe())

  #тут коли буду вміти візуазізувати ще додам графік, і буде на графіку красивіше та зрозуміліше персантиль

count    328.00
mean       7.07
std        1.70
min        4.00
25%        5.60
50%        7.10
75%        8.50
max       10.00
Name: Rating, dtype: float64

**Висновок до першого питання:**

Ми бачимо, що рейтинг в місті `Naypyitaw` середнім значенням має - 7.07 бала. Це достатньо високий рейтинг.

Також з `персантилю` (та майбутнього графіку) ми бачимо, що половина всіх відгуків на цьому значенні та вище. А значить тільки половина клієнтів задоволені магазинами(ставлять 7+балів).

Нам потрібно провести додаткове опитування критеріїв, які подобаються(не подобаються) клієнтам.
Потім на основі цих даних запровадити зміни, щоб відсоток середньої та вище оцінки сягав більшого значення.

**2 питання**

Де найнижча `"cogs"`?

In [8]:
df.loc[df['cogs'].idxmin()]

Invoice ID                       784-21-9238
Branch                                     C
City                               Naypyitaw
Customer type                         Member
Gender                                  Male
Product line               Sports and travel
Unit price                             10.17
Quantity                                   1
Tax 5%                                0.5085
Total                                10.6785
Date                                2/7/2019
Time                                   14:15
Payment                                 Cash
cogs                                   10.17
gross margin percentage             4.761905
gross income                          0.5085
Rating                                   5.9
Name: 822, dtype: object

**Висновок до другого питання:**

Ми бачимо, що найменший `cogs` обійшовся в 10.17. При цьому якщо порівняти `gross margin percentage` то він всюди однаковий.

Якщо говорити про відкриття нових магазині ще, то думаю варто відкривати аналоги з найменшим `cogs`, щоб менше грошей заморожувати в бізнесі. При цьому отримувати такий же `gross margin percentage`.



**3 питання**

Якого `"Customer type"` більше в місті `'Naypyitaw'`?

In [ ]:
df[df['City'] == 'Naypyitaw']['Customer type'].value_counts().idxmax()

'Member'

In [ ]:
df[df['City'] == 'Naypyitaw']['Customer type'].value_counts(normalize = True)

Customer type
Member    0.515244
Normal    0.484756
Name: proportion, dtype: float64

**Висновок на третє питання:**

Ми бачимо, що у місті `Naypyitaw` користувачів `Member` більше, ніж `Normal `. Проте ця різниця в кількості невелика.

Можна додатково дослідити чи роблять клієнти `Member `більші покупки за чеком, чи ні.

Якщо так, то тоді варто переводити `Normal `клієнтів через різні маркетингові інструменти в інший тип.
